autoload/magit/utils.vim

```diff
@@ -20,12 +20,43 @@ function! magit#utils#is_submodule(dirname)
 	return ( index(s:submodule_list, a:dirname) != -1 )
 endfunction
 
-" s:magit_cd_cmd: plugin variable to choose lcd/cd command, 'lcd' if exists,
-" 'cd' otherwise
-let s:magit_cd_cmd = exists('*haslocaldir') && haslocaldir() ? 'lcd ' : 'cd '
-" magit#utils#lcd: helper function to lcd. use cd if lcd doesn't exists
-function! magit#utils#lcd(dir)
-	execute s:magit_cd_cmd . a:dir
+" magit#utils#chdir will change the directory respecting
+" local/tab-local/global directory settings.
+function! magit#utils#chdir(dir)
+  " This is a dirty hack to fix tcd breakages on neovim.
+  " Future work should be based on nvim API and has a `if has('nvim')` check.
+  if has('nvim')
+    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
+  else
+    let chdir = exists('*haslocaldir') && haslocaldir() ? 'lcd' : 'cd'
```

Comment:
We can externalize both checks, but that would only make a difference if we define a function for each case. Otherwise we're swapping a strong check (i.e. `has('nvim')`) that won't change during runtime to something that can be overriden (i.e. `g:magit_running_on_neovim=1`).

By conditioning the implementation like that, we end up with an unmaintainable version of it:
```vim
if has('nvim')
  function! magit#utils#chdir(dir)
    let chdir = haslocaldir() ? 'lcd' : haslocaldir(-1, 0) ? 'tcd' : 'cd'
    exec chdir a:dir
  endfunction
elseif exists('*haslocaldir')
function! magit#utils#chdir(dir)
    let chdir = haslocaldir() ? 'lcd' : 'cd'
    exec chdir a:dir
  endfunction
else
function! magit#utils#chdir(dir)
    cd a:dir
  endfunction
endif
```